crates/ruff_python_formatter/Docs.md

```diff
@@ -1,5 +1,156 @@
+            item1.format()
+            text(","),
+            soft_line_break_or_space(),
+            item2.format(),
+            if_group_breaks(&text(","))
+        ]),
+        text(")")
+    ])]
+)
+```
+
+If everything fits on a single line, the group doesn't break and we get something like `("a", "b")`.
+If it doesn't, we get something like
+
+```python
+(
+    "a",
+    "b",
+)
+```
+
+For a list of expression, you don't need to format it manually but can use the `JoinBuilder` util,
+accessible through `.join_with`. Finish will write to the formatter internally.
+
+```rust
+f.join_with(&format_args!(text(","), soft_line_break_or_space()))
+    .entries(self.elts.iter().formatted())
+    .finish()?;
+// Here we need a trailing comma on the last entry of an expanded group since we have more
+// than one element
+write!(f, [if_group_breaks(&text(","))])
+```
+
+If you need avoid second mutable borrows with a builder, you can use `format_with(|f| { ... })` as
+a formattable element similar to `text()` or `group()`.
+
+The generic comment formatting in `FormatNodeRule` handles comments correctly for most nodes, e.g.
+preceding and end-of-line comments depending on the node range. Sometimes however, you may have
+dangling comments that are not before or after a node but inside of it, e.g.
+
+```python
+[
+    # here we use an empty list
+]
+```
+
+Here, you have to call `dangling_comments` manually and stubbing out `fmt_dangling_comments` in list
+formatting.
+
+```rust
+impl FormatNodeRule<ExprList> for FormatExprList {
+    fn fmt_fields(&self, item: &ExprList, f: &mut PyFormatter) -> FormatResult<()> {
+        // ...
+
+        write!(
+            f,
+            [group(&format_args![
+                text("["),
+                dangling_comments(dangling),
+                soft_block_indent(&items),
+                text("]")
+            ])]
+        )
+    }
+
+    fn fmt_dangling_comments(&self, _node: &ExprList, _f: &mut PyFormatter) -> FormatResult<()> {
+        // Handled as part of `fmt_fields`
+        Ok(())
+    }
+}
+```
+
+Comments are categorized into `Leading`, `Trailing` and `Dangling`, you can override this in
+`place_comment`.
+
+## Development notes
+
+Handling parentheses and comments are two major challenges in a python formatter.
+
+We have copied the majority of tests over from black and use [insta](https://insta.rs/docs/cli/) for
+snapshot testing with the diff between ruff and black, black output and ruff output. We put
+additional test cases in `resources/test/fixtures/ruff`.
+
+The full ruff test suite is slow, `cargo test -p ruff_python_formatter` is a lot faster.
+
+There is a `ruff_python_formatter` binary that avoid building and linking the main `ruff` crate.
+
+You can use `scratch.py` as a playground, e.g.
+`cargo run --bin ruff_python_formatter -- --emit stdout scratch.py`, which additional `--print-ir`
+and `--print-comments` options.
+
+## The orphan rules and out trait structure
```

Comment:
Is "out trait" a typo or am I misunderstanding the title?

Reply:
yep that's a type